AI AttackOrFleeFuzzy improvements

reaperrr opened the pull request:
I needed a break from the other stuff I'm working on, so I gave this a look.

The idea behind the string shortenings is the assumption that the FuzzyLibrary must be doing *some* kind of string comparison, so shorter strings should speed things up a little (likely insignificant, but who knows...).

The other commits should, at least in theory, make the AI behave a little smarter when making attack-or-flee decisions.

reaperrr committed:
Make AttackOrFleeFuzzy consider all DamageWarheads instead of only the first

reaperrr committed:
Factor number of bursts plus ReloadDelay into AttackOrFleeFuzzy.RelativePower

reaperrr committed:
Remove some redundancy in AttackOrFleeFuzzy

reaperrr commented:
Minor update to 2nd commit, pulled the .Clamp out of the foreach and added a comment detailing the reasons for the clamp values.

GraionDilach reviewed (approved):
:+1:

pchote commented on OpenRA.Mods.Common/AI/AttackOrFleeFuzzy.cs:
```diff
@@ -221,8 +221,8 @@ static float RelativePower(IEnumerable<Actor> own, IEnumerable<Actor> enemy)
 				{
-					var warhead = arm.Weapon.Warheads.OfType<DamageWarhead>().FirstOrDefault();
-					if (warhead != null)
+					var dmgWarheads = arm.Weapon.Warheads.OfType<DamageWarhead>();
```
`damageWarheads` please.

pchote commented on OpenRA.Mods.Common/AI/AttackOrFleeFuzzy.cs:
```diff
@@ -25,7 +25,7 @@ sealed class AttackOrFleeFuzzy
 			"if ((OwnHealth is Normal) " +
 			"and ((EnemyHealth is NearDead) or (EnemyHealth is Injured) or (EnemyHealth is Normal)) " +
-			"and ((RelativeAttackPower is Weak) or (RelativeAttackPower is Equal) or (RelativeAttackPower is Strong)) " +
+			"and ((RelAtkPwr is Weak) or (RelAtkPwr is Equal) or (RelAtkPwr is Strong)) " +
```
I don't buy that these would make any measurable performance difference, and IMO the shortened versions are a lot less readable.  Can we please drop these renames?

pchote commented on OpenRA.Mods.Common/AI/AttackOrFleeFuzzy.cs:
```diff
@@ -23,104 +23,92 @@ sealed class AttackOrFleeFuzzy
-			"and ((RelativeSpeed is Slow) or (RelativeSpeed is Equal) or (RelativeSpeed is Fast))) " +
+			"if ((OwnHP is Normal) " +
+			"and ((EnemyHP is NearDead) or (EnemyHP is Injured) or (EnemyHP is Normal)) " +
+			"and ((RelAtkPwr is Weak) or (RelAtkPwr is Equal) or (RelAtkPwr is Strong)) " +
```
I don't buy that these would make any measurable performance difference, and IMO the shortened versions are a lot less readable.  Can we please drop these renames?

pchote commented on OpenRA.Mods.Common/AI/AttackOrFleeFuzzy.cs:
```diff
@@ -221,9 +209,16 @@ static float RelativePower(IEnumerable<Actor> own, IEnumerable<Actor> enemy)
+					// High dmg/low ROF weapons shouldn't be rated too low as high dmg/shot can outweigh mere dps due to likelier 1-hit-kills.
+					// TODO: Revisit this at some point to replace the arbitrary cap with something smarter.
+					var totalReloadDelay = arm.Weapon.ReloadDelay + (arm.Weapon.BurstDelays[0] * (burst - 1)).Clamp(1, 200);
+					var dmgWarheads = arm.Weapon.Warheads.OfType<DamageWarhead>();
```
`damageWarheads` please.

reaperrr committed:
Make sure AttackOrFlee returns Attack when it makes sense

Not listing enemy "NearDead" and own relative atk power "Strong" here looked wrong, and might've resulted in no decision being made.

reaperrr commented:
Updated.

pchote reviewed (approved):
Code changes look sensible and I didn't notice any obvious regressions ingame.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/fec14d143977fcd44623c003cf6cbf21210a2baf
